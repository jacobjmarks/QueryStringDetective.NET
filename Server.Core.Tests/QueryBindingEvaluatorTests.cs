namespace Server.Core.Tests;

public class QueryBindingEvaluatorTests
{
    public static readonly TheoryData<string> TestCases = new()
    {
        "c",
        "foo",
        "a&q=b&q=c",
        "null",
        "0",
        "1",
        "-1",
        "true",
        "false",
        "yes",
        "no",
        "1&q=2&q=3",
        "1&q=2&q=.3",
        "-1&q=0&q=1",
        $"{sbyte.MinValue}",
        $"{sbyte.MaxValue}",
        $"{byte.MaxValue}",
        $"{short.MinValue}",
        $"{short.MaxValue}",
        $"{ushort.MaxValue}",
        $"{int.MinValue}",
        $"{int.MaxValue}",
        $"{long.MinValue}",
        $"{long.MaxValue}",
        $"{ulong.MaxValue}",
        Uri.EscapeDataString($"{double.MinValue}"),
        Uri.EscapeDataString($"{double.MaxValue}"),
        $"{decimal.MinValue}",
        $"{decimal.MaxValue}",
        Uri.EscapeDataString($"{float.MinValue}"),
        Uri.EscapeDataString($"{float.MaxValue}"),
    };

    [Theory]
    [MemberData(nameof(TestCases))]
    public async Task AssertBinding(string queryString)
    {
        using var bindingEvaluator = new QueryBindingEvaluator();
        var results = await bindingEvaluator.EvaluateAsync("?q=" + queryString);

        var content = $"Input: {queryString}"
            + "\n\nBindings:\n";

        foreach (var result in results)
        {
            content += "\n" + result.Type;
            content += "\n  " + string.Join("\n  ", result.Results
                .Select(r => $"{r.Key}\n    {(
                    r.Value.IsErroneous
                    ? $"e = {r.Value.Error.Message} ({r.Value.Error.Detail})"
                    : $"r = {r.Value.Result}")}"));

            content += "\n";
        }

        content.Should().MatchSnapshot(SnapshotNameExtension.Create(ToValidFileName(queryString)));
    }

    private static string ToValidFileName(string value)
    {
        foreach (var invalidChar in Path.GetInvalidFileNameChars())
            value = value.Replace(invalidChar, '_');

        return value;
    }
}