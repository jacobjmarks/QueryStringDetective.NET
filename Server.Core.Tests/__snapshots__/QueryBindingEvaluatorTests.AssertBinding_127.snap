Input: 127

Bindings:

string
  MinimalApis
    r = "127"

string[]
  MinimalApis
    r = ["127"]

char
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char q" from "127".)

char[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char[] q" from "127".)

int
  MinimalApis
    r = 127

int[]
  MinimalApis
    r = [127]

double
  MinimalApis
    r = 127

double[]
  MinimalApis
    r = [127]

float
  MinimalApis
    r = 127

float[]
  MinimalApis
    r = [127]

decimal
  MinimalApis
    r = 127

decimal[]
  MinimalApis
    r = [127]

bool
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool q" from "127".)

bool[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool[] q" from "127".)

uint
  MinimalApis
    r = 127

uint[]
  MinimalApis
    r = [127]

long
  MinimalApis
    r = 127

long[]
  MinimalApis
    r = [127]

ulong
  MinimalApis
    r = 127

ulong[]
  MinimalApis
    r = [127]

sbyte
  MinimalApis
    r = 127

sbyte[]
  MinimalApis
    r = [127]

short
  MinimalApis
    r = 127

short[]
  MinimalApis
    r = [127]

ushort
  MinimalApis
    r = 127

ushort[]
  MinimalApis
    r = [127]

byte
  MinimalApis
    r = 127

byte[]
  MinimalApis
    r = "fw=="
